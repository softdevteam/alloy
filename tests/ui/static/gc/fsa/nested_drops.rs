#![feature(gc)]
#![feature(negative_impls)]
#![feature(rustc_private)]
#![allow(dead_code)]
#![allow(unused_variables)]
include!{"./auxiliary/types.rs"}


extern crate libc;

struct HasSafeDrop;
struct HasUnsafeDrop1;
struct HasUnsafeDrop2;

impl Drop for HasSafeDrop { fn drop(&mut self) { let x = 1;} }

impl Drop for HasUnsafeDrop1 {
    fn drop(&mut self) {
        unsafe { libc::malloc(8) as *mut i32 };
        foo();
    }
}

impl Drop for HasUnsafeDrop2 {
    fn drop(&mut self) {
        unsafe { libc::calloc(8, 8) as *mut i32 };
    }
}

fn foo() {
    let s = HasUnsafeDrop2;
}

#[derive(Debug)]
struct HasUnsafeNestedDrop(u8);
#[derive(Debug)]
struct HasSafeNestedDrop(u8);

impl Drop for HasUnsafeNestedDrop  {
    fn drop(&mut self) { let s = HasUnsafeDrop1; }
}

impl Drop for HasSafeNestedDrop  {
    fn drop(&mut self) { let s = HasSafeDrop; }
}

fn main() {
    Gc::new(FinalizerUnsafeWrapper(HasUnsafeNestedDrop(1)));
    //~^ ERROR: The drop method for `HasUnsafeNestedDrop` cannot be safely finalized.
    //~| ERROR: The drop method for `HasUnsafeNestedDrop` cannot be safely finalized.

    Gc::new(FinalizerUnsafeWrapper(HasSafeNestedDrop(1)));
}
