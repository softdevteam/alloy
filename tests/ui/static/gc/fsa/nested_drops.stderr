error: The drop method for `HasUnsafeNestedDrop` cannot be safely finalized.
  --> $DIR/nested_drops.rs:48:13
   |
LL |         unsafe { libc::malloc(8) as *mut i32 };
   |                  --------------- this function call may be unsafe to use in a finalizer.
...
LL |     Gc::new(FinalizerUnsafeWrapper(HasUnsafeNestedDrop(1)));
   |     --------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<FinalizerUnsafeWrapper<HasUnsafeNestedDrop>>` here.

error: The drop method for `HasUnsafeNestedDrop` cannot be safely finalized.
  --> $DIR/nested_drops.rs:48:13
   |
LL |         unsafe { libc::calloc(8, 8) as *mut i32 };
   |                  ------------------ this function call may be unsafe to use in a finalizer.
...
LL |     Gc::new(FinalizerUnsafeWrapper(HasUnsafeNestedDrop(1)));
   |     --------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<FinalizerUnsafeWrapper<HasUnsafeNestedDrop>>` here.

error: aborting due to 2 previous errors

